### PR TITLE
Da zi ppo

### PR DESCRIPTION
**Pull request by Dalouvid**
#### Summary of additions and changes

* Implemented the PPO 
* Implemented GAE 

Problem with GAE is that the number of Rewards should not be the same as the number of states. 
When I tried to test that, the test failed.. but in deployment no error was raised... therefore I think something is wrong. 
Couldn't figure out what's wrong yet. Guess something with the axis. 
#### How to test this pull request

```

```

### CODE REVIEW

File: swarmrl/networks/flax_network.py [hunk 1]
```diff
@@ -133,8 +133,7 @@ def compute_action(self, feature_vector: np.ndarray, explore_mode: bool = False)
 
         if explore_mode:
             index = self.exploration_policy(index, len(model_output))
-
-        return index, model_output[index]
+        return index, jax.nn.softmax(model_output)[index]
```
Comment (SamTov) [inline swarmrl/networks/flax_network.py:136]: This will break the normal loss right?

File: swarmrl/tasks/searching/gradient_sensing.py [hunk 2]
```diff
@@ -88,11 +94,16 @@ def __call__(self, observable: np.ndarray):
...
+                Observable of a single colloid.
+        colloid : object
+                The colloid, the reward should be computed for.
+        colloids : list
+                A list of all colloids in the system.
+        other_colloids : list
+                A list of all other colloids in the system.
         Returns
-        -------
+        ----------
```
Comment (SamTov) [inline swarmrl/tasks/searching/gradient_sensing.py:106]: Why the additional lines?

**Review by SamTov — CHANGES_REQUESTED** [review]
Just a small review of things I notices quickly, I am still looking into the shape error.

**Comment by SamTov** [comment]
@Dalouvid I have managed to reproduce the error in the main code, it is just wrong. The reason nothing came up is because a compiled device array was being passed to the function. It seems that when n + 1 is called as an index on compiled arrays. it simply returns a value from elsewhere in the array.

File: swarmrl/networks/flax_network.py (hunk 1 above)
Comment (Dalouvid) [reply swarmrl/networks/flax_network.py:136]: You mean it would break the Policy Gradient Loss? 
As far as I can see, no. 
In policy gradient loss we don't use this information. We only give it the chosen action.

File: swarmrl/tasks/searching/gradient_sensing.py (hunk 2 above)
Comment (Dalouvid) [reply swarmrl/tasks/searching/gradient_sensing.py:106]: I think that was by accident :D
